app/src/main/java/net/gsantner/markor/ui/SearchReplaceDialog.java

```diff
@@ -0,0 +1,316 @@
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+import java.util.regex.PatternSyntaxException;
+
+public class SearchReplaceDialog {
+
+    private static final String SEARCH_REPLACE_SETTINGS = "search_replace_dialog_settings";
+    private static final String RECENT_SEARCH_REPLACE_STRING = "recent_search_replace";
+    private static final int MAX_RECENT_SEARCH_REPLACE = 10;
+
+    private final EditText searchText;
+    private final EditText replaceText;
+    private final CheckBox regexCheckBox;
+    private final CheckBox multilineCheckBox;
+    private final TextView matchState;
+    private final Button replaceFirst;
+    private final Button replaceAll;
+
+    private final Activity _activity;
+    private final TextView _text;
+
+    private final int[] sel;
+    private final CharSequence region;
+
+    private final List<ReplaceGroup> recentReplaces;
+
+    public static void showSearchReplaceDialog(final Activity activity, final TextView text) {
+        new SearchReplaceDialog(activity, text);
+    }
+
+    private SearchReplaceDialog(final Activity activity, final TextView text) {
+
+        _activity = activity;
+        _text = text;
+
+        final Resources res = activity.getResources();
+        final AlertDialog.Builder builder = new AlertDialog.Builder(activity);
+        final View viewRoot = activity.getLayoutInflater().inflate(R.layout.search_replace_dialog, null);
+        final AtomicReference<Dialog> dialog = new AtomicReference<>();
+
+        searchText = viewRoot.findViewById(R.id.search_input);
+        replaceText = viewRoot.findViewById(R.id.replace_input);
+        regexCheckBox = viewRoot.findViewById(R.id.use_regex);
+        multilineCheckBox = viewRoot.findViewById(R.id.multiline);
+        matchState = viewRoot.findViewById(R.id.match_count_or_error);
+        replaceFirst = viewRoot.findViewById(R.id.replace_first);
+        replaceAll = viewRoot.findViewById(R.id.replace_all);
+
+        recentReplaces = loadRecentReplaces();
+
+        // Set region for replace
+        sel = StringUtils.getSelection(_text);
+        // no selection, replace with all
+        if (sel[0] == sel[1]) {
+            sel[0] = 0;
+            sel[1] = text.length();
+        }
+        region = _text.getText().subSequence(sel[0], sel[1]);
+
+        final ListPopupWindow popupWindow = new ListPopupWindow(activity);
+
+        // Popup window for ComboBox
+        popupWindow.setAdapter(new ArrayAdapter<ReplaceGroup>(activity, android.R.layout.simple_list_item_1, recentReplaces) {
+            @NonNull
+            @Override
+            public View getView(int pos, @Nullable View view, @NonNull ViewGroup parent) {
+                final TextView textView = (TextView) super.getView(pos, view, parent);
+
+                final ReplaceGroup rg = getItem(pos);
+                textView.setText(res.getString(R.string.search_replace_recent_format, rg._search, rg._replace, rg._isRegex, rg._isMultiline));
```

Comment:
This is ugliest part of this dialog. Each line of the recents pattern is simply:

```
Search for: <search pattern>
Replace with: <replace pattern>
Use regex: <true/false>, Multiline regex <true/false>

```
(note the empty line). It works ok, but is pretty ugly. Any suggestions are most welcome